File: packages/typescript/src/portable.ts
/** Portable YAML input shared by artifact and schema reads.
 *
 * The input, node, and scalar ceilings were removed. They only answered whether a
 * hostile document could exhaust the parser, and softschema reads artifacts its own
 * callers just wrote. The portable-value rules here stay, because they are what makes a
 * document mean the same thing in both runtimes.
 */
import { readFileSync } from "node:fs";
import { writeFileSync } from "atomically";
import { isAlias, isCollection, isPair, isScalar, parseDocument, visit } from "yaml";

/** Largest integer that survives a round trip through a JS number. */
const MAX_SAFE_INTEGER = 9_007_199_254_740_991;

/** Simultaneously open collections, including the root.
 *
 * A portability rule rather than a resource guard, and the reason it outlived the three
 * ceilings above. Left to the host, V8 parses past depth 10,000 while CPython's default
 * recursion limit stops the Python constructor around 491, so any document between those
 * bounds would be valid here and a crash there. Keep this value identical to `MAX_DEPTH`
 * in the Python `_portable` module; the shared depth vectors check that they agree.
 */
const MAX_DEPTH = 64;

export class PortableInputError extends Error {
  constructor(
    readonly code: string,
    message: string,
  ) {
    super(message);
  }
}

/** A leading byte order mark, which decoding removes rather than keeps as a character. */
const BOM = "\uFEFF";

/**
 * Read one UTF-8 artifact, dropping a leading byte order mark.
 *
 * The BOM has to go, and it has to go here, because this is the one function both runtimes
 * route every artifact and schema read through. `TextDecoder`'s default `ignoreBOM: false`
 * already strips it, so the explicit check below is a no-op today; it is written out
 * anyway, the same way Python's `read_utf8` writes it out, because that default is the
 * only thing holding the two runtimes together and it is invisible at the call site.
 * Swapping in `readFileSync(path, "utf8")` — which keeps the mark — would silently
 * reintroduce the divergence this replaced.
 *
 * That divergence was real: the fence scan and the reader both compare a first line
 * against `---`, and `"\uFEFF---"` is not `"---"`, so Python called a BOM-prefixed
 * artifact fenceless while TypeScript read it without complaint. Reporting that document
 * as having no frontmatter is the diagnostic this codebase exists to avoid — the block is
 * plainly there. Only position zero is examined; a U+FEFF anywhere else is a real
 * character and survives.
 */
export function readUtf8(path: string): string {
  let text: string;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(readFileSync(path));
  } catch (error) {
    if (error instanceof TypeError) {
      throw new PortableInputError("invalid_utf8", "input is not valid UTF-8");
    }
    throw error;
  }
  return text.startsWith(BOM) ? text.slice(1) : text;
}

export function parsePortableYaml(text: string): unknown {
  let document: ReturnType<typeof parseDocument>;
  try {
    document = parseDocument(text, { uniqueKeys: true });
  } catch (error) {
    if (error instanceof RangeError || String(error).includes("Maximum call stack size exceeded")) {
      throw new PortableInputError("yaml_limit", "YAML exceeds the depth limit");
    }
    throw error;
  }
  if (document.errors.length > 0) {
    const message = document.errors[0]?.message ?? "invalid YAML";
    if (message.includes("Maximum call stack size exceeded")) {
      throw new PortableInputError("yaml_limit", "YAML exceeds the depth limit");
    }
    const code = message.includes("Map keys must be unique")
      ? "yaml_duplicate_key"
      : "yaml_parse_error";
    throw new PortableInputError(code, message);
  }
  if (document.warnings.length > 0) {
    throw new PortableInputError(
      "yaml_custom_tag",
      document.warnings[0]?.message ?? "unsupported tag",
    );
  }

  let hasAlias = false;
  visit(document, (_key, node, path) => {
    // `path` is bounded by MAX_DEPTH, because exceeding it throws on the next node.
    const depth = path.reduce((total, ancestor) => total + (isCollection(ancestor) ? 1 : 0), 0);
    if (depth > MAX_DEPTH) {
      throw new PortableInputError("yaml_limit", "YAML exceeds the depth limit");
    }
    if (isPair(node)) {
      if (!isScalar(node.key) || typeof node.key.value !== "string") {
        throw new PortableInputError("yaml_non_string_key", "mapping keys must be strings");
      }
      if (node.key.value === "<<") {
        throw new PortableInputError("yaml_merge_key", "YAML merge keys are not supported");
      }
      return;
    }
    if (
      isAlias(node) ||
      (typeof node === "object" &&
        node !== null &&
        "anchor" in node &&
        typeof node.anchor === "string")
    ) {
      hasAlias = true;
    }
    if (isScalar(node) && node.tag !== undefined) {
      throw new PortableInputError("yaml_custom_tag", "explicit YAML tags are not supported");
    }
    if (
      isScalar(node) &&
      typeof node.value === "number" &&
      Number.isInteger(node.value) &&
      Math.abs(node.value) > MAX_SAFE_INTEGER &&
      typeof node.source === "string" &&
      !/[.eE]/u.test(node.source)
    ) {
      throw new PortableInputError("number_out_of_range", "integer exceeds the safe range");
    }
  });
  if (hasAlias) {
    throw new PortableInputError("yaml_alias", "YAML aliases and anchors are not supported");
  }

  const value = document.toJS();
  checkPortableValue(value);
  return value;
}

export function checkPortableValue(root: unknown): void {
  const stack: unknown[] = [root];
  while (stack.length > 0) {
    const value = stack.pop();
    if (value === null || typeof value === "boolean") continue;
    if (typeof value === "string") {
      for (let index = 0; index < value.length; index += 1) {
        const code = value.charCodeAt(index);
        const next = value.charCodeAt(index + 1);
        if (code >= 0xd800 && code <= 0xdbff && next >= 0xdc00 && next <= 0xdfff) {
          index += 1;
        } else if (code >= 0xd800 && code <= 0xdfff) {
          throw new PortableInputError(
            "yaml_unsupported_scalar",
            "lone surrogate is not supported",
          );
        }
      }
      continue;
    }
    if (typeof value === "number") {
      if (!Number.isFinite(value)) {
        throw new PortableInputError("number_out_of_range", "number must be finite");
      }
      if (Object.is(value, -0)) {
        throw new PortableInputError("number_negative_zero", "negative zero is not supported");
      }
      continue;
    }
    if (Array.isArray(value)) {
      stack.push(...value);
      continue;
    }
    if (typeof value === "object") {
      const prototype = Object.getPrototypeOf(value);
      if (prototype !== Object.prototype && prototype !== null) {
        throw new PortableInputError(
          "yaml_unsupported_scalar",
          `host-native ${value.constructor?.name ?? "object"} values are not portable; use JSON-compatible values`,
        );
      }
      stack.push(...Object.values(value));
      continue;
    }
    throw new PortableInputError(
      "yaml_unsupported_scalar",
      `unsupported YAML value: ${typeof value}`,
    );
  }
}

/**
 * Where a frontmatter-md document's metadata sits inside its text.
 *
 * All three offsets index into the *original* text, so a caller can put the document
 * back together exactly:
 *
 *     text.slice(0, metadataOffset) + newMetadata +
 *       text.slice(metadataEnd, bodyOffset) + text.slice(bodyOffset)
 *
 * That middle slice is the closing fence, kept verbatim rather than re-synthesized. This
 * is the difference from `readFrontmatterDoc`, which splits on `/\r?\n/` and rejoins
 * with `"\n"`: fine for reading values, lossy for writing the file back.
 */
export interface FrontmatterSplit {
  metadataText: string;
  metadataOffset: number;
  metadataEnd: number;
  bodyOffset: number;
}

/**
 * Split a frontmatter-md document without disturbing its body.
 *
 * Returns `null` when there is no frontmatter *region to rewrite*: no leading fence, an
 * unterminated fence, or an empty block whose end fence is the very next line.
 *
 * That is narrower than it looks. Of those three, `readFrontmatterDoc` agrees on two —
 * no leading fence and an empty block — but an unterminated fence is a *reader error*
 * there, not a fenceless document. So `null` from this function must never be read as
 * "this document has no frontmatter fence"; use `opensFrontmatterFence` for that.
 * Reading `null` as fenceless is what once let profile detection route an
 * unterminated-fence document to pure-yaml and call it valid while the reader refused to
 * open it at all.
 *
 * This is the same hand-rolled scan as Python's `split_frontmatter`, line for line, and
 * deliberately so. The fence rules have to stay identical to the reader's — if the two
 * disagree about where the frontmatter ends, a repair pass writes one region and
 * validation reads another — and writing the scan twice is what keeps the two runtimes
 * splitting identically.
 */
export function splitFrontmatter(text: string): FrontmatterSplit | null {
  // Scan by offset rather than by `split()`, because the offsets are the whole point: a
  // `\r\n` document must keep its `\r\n` body byte-for-byte.
  const first = lineEnd(text, 0);
  if (first === null || text.slice(0, first.contentEnd).trimEnd() !== "---") return null;
  const metadataOffset = first.next;
  let cursor = metadataOffset;
  while (cursor < text.length) {
    const line = lineEnd(text, cursor);
    if (line === null) break; // unreachable: the loop guard guarantees a line remains
    if (text.slice(cursor, line.contentEnd).trimEnd() === "---") {
      const metadataText = text.slice(metadataOffset, cursor);
      // An empty block (end fence on the very next line) is the portable
      // no_frontmatter case, the same as the reader's `end === 1`.
      if (metadataText.trim() === "") return null;
      return { metadataText, metadataOffset, metadataEnd: cursor, bodyOffset: line.next };
    }
    cursor = line.next;
  }
  return null; // unterminated fence: no region to rewrite (still a reader error)
}

/**
 * Whether the document's first line is a frontmatter opening fence.
 *
 * Profile detection needs this and cannot use `splitFrontmatter`, which returns `null`
 * both for a document that never opened a fence and for one that opened a fence and
 * never closed it. Those two get opposite treatment: the first may be a pure-yaml
 * artifact, while the second is a frontmatter-md document the reader will reject.
 * Detection has to tell them apart before it can decide a profile, and without parsing,
 * because an artifact awaiting repair does not parse.
 *
 * Only the opening fence is examined; a document that opens one is frontmatter-md
 * whether or not it closes it, exactly as `readFrontmatterDoc` treats it.
 */
export function opensFrontmatterFence(text: string): boolean {
  const first = lineEnd(text, 0);
  return first !== null && text.slice(0, first.contentEnd).trimEnd() === "---";
}

/**
 * The content end (before the line break) and the start of the next line.
 *
 * A final line with no trailing newline is a line. Its content ends at EOF, and the
 * "next line" starts there too, so a caller that keeps scanning stops on the next
 * iteration.
 *
 * That has to be so, because `readFrontmatterDoc` splits with `split(/\r?\n/)` and
 * Python's `read_frontmatter_doc` with `splitlines()`, and both keep such a line. A scan
 * that dropped it would disagree with the reader about where a document's fences are,
 * which is the disagreement this module exists to prevent. Concretely, it made
 * `splitFrontmatter` report "no region to rewrite" for a document ending exactly at its
 * closing fence, so `--repair` silently skipped an artifact it could fix, and it made
 * `opensFrontmatterFence` call a lone `---` fenceless.
 *
 * Returns `null` only when `from` is at or past the end, where no line remains.
 */
function lineEnd(text: string, from: number): { contentEnd: number; next: number } | null {
  if (from >= text.length) return null;
  const index = text.indexOf("\n", from);
  if (index === -1) return { contentEnd: text.length, next: text.length };
  return { contentEnd: index, next: index + 1 };
}

/**
 * Parse YAML in round-trip mode: the document keeps how the author wrote it.
 *
 * Round-trip parsing is what lets a one-scalar correction stay a one-scalar diff. The
 * `yaml` package's `parseDocument` retains quotes, comments, key order, and line
 * structure, which is the direct analogue of ruamel's `typ="rt"` on the Python side.
 */
export function parseRoundTrip(text: string): ReturnType<typeof parseDocument> {
  return parseDocument(text, { uniqueKeys: true });
}

/**
 * Serialize a round-trip document with the emitter settings Python's `round_trip_yaml`
 * uses, so both runtimes write the same bytes for the same edit.
 *
 * `lineWidth: 0` disables re-wrapping (ruamel's `width = 4096` in practice), and
 * `nullStr: "null"` pins the spelling ruamel is pinned to — left alone, both emitters
 * make it depend on position in the document.
 *
 * `singleQuote: true` earns its place twice. It is what makes a newly quoted scalar come
 * out `'1850'` here as it does in ruamel, rather than `"1850"`, which is a byte-parity
 * break in the one value this pass exists to write. It also stops the emitter rewriting
 * an author's `'single'` into `"double"` — a whole-file restyling diff around a
 * one-scalar fix. Without it both regressions appear together, and only the second one
 * is visible in a golden transcript.
 */
export function dumpRoundTrip(document: ReturnType<typeof parseDocument>): string {
  return document.toString({
    lineWidth: 0,
    indent: 2,
    indentSeq: true,
    nullStr: "null",
    singleQuote: true,
  });
}

/**
 * Write an artifact back: atomically, and without touching its line endings.
 *
 * The one write path for every pass that rewrites an artifact (repair, conform, the
 * combined pipeline), so the "atomic, endings untouched" contract has a single home. The
 * text carries exactly the endings the caller preserved.
 */
export function writeArtifactText(path: string, text: string): void {
  writeFileSync(path, text, { encoding: "utf8" });
}

File: packages/python/src/softschema/_portable.py
"""Bounded UTF-8 and portable YAML input shared by artifact and schema reads."""

from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import date
from io import StringIO
from pathlib import Path
from typing import Any

from ruamel.yaml import YAML, YAMLError
from ruamel.yaml.constructor import SafeConstructor
from ruamel.yaml.events import (
    AliasEvent,
    MappingEndEvent,
    MappingStartEvent,
    ScalarEvent,
    SequenceEndEvent,
    SequenceStartEvent,
)
from ruamel.yaml.nodes import ScalarNode

MAX_SAFE_INTEGER = 9_007_199_254_740_991
"""Largest integer that survives a round trip through a JS number."""

MAX_DEPTH = 64
"""Simultaneously open collections, including the root.

A portability rule, not a resource guard, which is why it survives while the input,
node, and scalar ceilings did not. Those three only answered whether a hostile document
could exhaust the parser, and softschema reads artifacts its own callers just wrote.
Depth answers a different question: how deep a document still means the same thing in
both runtimes.

Without an explicit rule the host stack decides, and the two hosts disagree by an order
of magnitude. CPython's default recursion limit of 1,000 lets this constructor reach
depth 491 from an empty caller stack — less from a real one, since the caller's own
frames come out of the same budget — while V8 parses past 10,000. A document between
those two bounds would be valid in TypeScript and a crash in Python, which is exactly
what the shared vectors exist to prevent.

64 is far above any real artifact and leaves roughly 7x headroom under the measured
ceiling, so the check below always fires before the recursion does. Raising it is safe
only while that headroom holds; past a few hundred the Python guard stops being reachable
and `RecursionError` takes over.
"""


def _construct_timestamp_as_string(_constructor: SafeConstructor, node: ScalarNode) -> str:
    """Keep implicit YAML timestamps inside the portable string domain."""
    return node.value


class _PortableConstructor(SafeConstructor):
    """Safe constructor overrides scoped to softschema parser instances."""


# The first subclass registration copies ruamel's inherited constructor registry.
_PortableConstructor.add_constructor("tag:yaml.org,2002:timestamp", _construct_timestamp_as_string)


class PortableInputError(ValueError):
    """Stable reason for input outside the portable value domain."""

    def __init__(self, code: str, message: str) -> None:
        self.code = code
        super().__init__(message)


def read_utf8(path: Path) -> str:
    """Read one UTF-8 artifact."""
    try:
        return path.read_bytes().decode("utf-8", errors="strict")
    except UnicodeDecodeError as exc:
        raise PortableInputError("invalid_utf8", "input is not valid UTF-8") from exc


def parse_yaml(text: str) -> Any:
    """Parse YAML with the existing library after a representation preflight."""
    yaml = YAML(typ="safe")
    yaml.Constructor = _PortableConstructor
    stack: list[tuple[str, bool]] = []
    has_alias = False

    def consume_parent_slot() -> None:
        if stack and stack[-1][0] == "map":
            kind, expects_key = stack[-1]
            stack[-1] = (kind, not expects_key)

    try:
        for event in yaml.parse(text):
            if isinstance(event, AliasEvent) or getattr(event, "anchor", None) is not None:
                has_alias = True
            if getattr(event, "tag", None) is not None:
                raise PortableInputError("yaml_custom_tag", "explicit YAML tags are not supported")
            if isinstance(event, ScalarEvent):
                if stack and stack[-1] == ("map", True) and event.value == "<<":
                    raise PortableInputError("yaml_merge_key", "YAML merge keys are not supported")
                consume_parent_slot()
            elif isinstance(event, (MappingStartEvent, SequenceStartEvent)):
                consume_parent_slot()
                stack.append(
                    ("map", True) if isinstance(event, MappingStartEvent) else ("seq", False)
                )
                if len(stack) > MAX_DEPTH:
                    raise PortableInputError("yaml_limit", "YAML nesting exceeds the depth limit")
            elif isinstance(event, (MappingEndEvent, SequenceEndEvent)):
                stack.pop()
        if has_alias:
            raise PortableInputError("yaml_alias", "YAML aliases and anchors are not supported")
        value = yaml.load(text)
    except PortableInputError:
        raise
    except RecursionError as exc:
        # The preflight above rejects anything past MAX_DEPTH, so construction should
        # never recurse this far. Mirror the TypeScript RangeError handler anyway: a
        # caller already deep in its own stack can exhaust the budget at a legal depth,
        # and that has to stay a structured result rather than escape validate_artifact.
        raise PortableInputError("yaml_limit", "YAML nesting exhausted the parser stack") from exc
    except YAMLError as exc:
        code = "yaml_duplicate_key" if "duplicate key" in str(exc).lower() else "yaml_parse_error"
        raise PortableInputError(code, str(exc)) from exc
    except ValueError as exc:
        raise PortableInputError("yaml_parse_error", str(exc)) from exc
    _check_value(value)
    return value


def _check_value(root: Any) -> None:
    stack: list[Any] = [root]
    while stack:
        value = stack.pop()
        if value is None or isinstance(value, bool):
            continue
        if isinstance(value, str):
            if any(0xD800 <= ord(char) <= 0xDFFF for char in value):
                raise PortableInputError(
                    "yaml_unsupported_scalar", "lone surrogate is not supported"
                )
            continue
        if isinstance(value, int):
            if abs(value) > MAX_SAFE_INTEGER:
                raise PortableInputError("number_out_of_range", "integer exceeds the safe range")
            continue
        if isinstance(value, float):
            if not math.isfinite(value):
                raise PortableInputError("number_out_of_range", "number must be finite")
            if value.hex() == "-0x0.0p+0":
                raise PortableInputError("number_negative_zero", "negative zero is not supported")
            continue
        if isinstance(value, date):
            raise PortableInputError(
                "yaml_unsupported_scalar",
                "host-native date and datetime values are not portable; use an ISO string",
            )
        if isinstance(value, list):
            stack.extend(value)
            continue
        if isinstance(value, dict):
            for key, item in value.items():
                if not isinstance(key, str):
                    raise PortableInputError("yaml_non_string_key", "mapping keys must be strings")
                stack.append(item)
            continue
        raise PortableInputError(
            "yaml_unsupported_scalar", f"unsupported YAML value: {type(value).__name__}"
        )


@dataclass(frozen=True)
class FrontmatterSplit:
    """Where a frontmatter-md document's metadata sits inside its text.

    All three fields are character offsets into the *original* text, so a caller can put
    the document back together exactly:

        text[:metadata_offset] + new_metadata + text[metadata_end:body_offset] + body

    That middle slice is the closing fence, kept verbatim rather than re-synthesized.
    This is the difference from :func:`read_frontmatter_doc`, which splits on
    ``splitlines()`` and rejoins with ``"\n"``: fine for reading values, lossy for
    writing the file back.
    """

    metadata_text: str
    metadata_offset: int
    body_offset: int

    @property
    def metadata_end(self) -> int:
        return self.metadata_offset + len(self.metadata_text)


def split_frontmatter(text: str) -> FrontmatterSplit | None:
    """Split a frontmatter-md document without disturbing its body.

    Returns ``None`` when the document has no frontmatter, matching what
    :func:`read_frontmatter_doc` reports: no leading fence, an unterminated fence, or an
    empty block whose end fence is the very next line.

    The scan is deliberately hand-rolled and byte-oriented rather than delegated to
    ``frontmatter_format``, for two reasons. The fence rules have to stay identical to
    that reader's — if the two disagree about where the frontmatter ends, a repair pass
    writes one region and validation reads another. And the TypeScript implementation has
    no equivalent package, so writing the same scan twice is what keeps the two runtimes
    splitting identically.
    """
    first = _line_end(text, 0)
    if first is None or text[: first[0]].rstrip() != "---":
        return None
    metadata_offset = first[1]
    cursor = metadata_offset
    while cursor < len(text):
        line = _line_end(text, cursor)
        if line is None:
            return None  # unterminated fence: no frontmatter to speak of
        if text[cursor : line[0]].rstrip() == "---":
            metadata_text = text[metadata_offset:cursor]
            # An empty block (end fence on the very next line) is the portable
            # no_frontmatter case, the same as the reader's ``end == 1``.
            if not metadata_text.strip():
                return None
            return FrontmatterSplit(
                metadata_text=metadata_text,
                metadata_offset=metadata_offset,
                body_offset=line[1],
            )
        cursor = line[1]
    return None


def _line_end(text: str, start: int) -> tuple[int, int] | None:
    """The content end (before the line break) and the start of the next line."""
    index = text.find("\n", start)
    if index == -1:
        return None
    return index, index + 1


def round_trip_yaml() -> YAML:
    """A ruamel round-trip parser/emitter that preserves how the author wrote things.

    Round-trip mode is what lets a one-scalar correction stay a one-scalar diff: quotes,
    comments, key order, and long lines survive because the loader hands back nodes that
    remember their own notation.

    ``allow_aliases`` is deliberately absent. metaproc's equivalent enables it to keep an
    author's ``*ref`` from expanding into a duplicated mapping, but softschema's
    :func:`parse_yaml` rejects aliases and anchors outright, so an artifact carrying one
    never reaches this emitter.

    The null representer is pinned anyway. ruamel spells a null as an empty value once
    any object has been represented and as ``null`` before that, which makes the spelling
    depend on position in the document. A null this pass did not touch changing shape is
    exactly the unasked-for diff round-trip mode exists to prevent.
    """
    yaml = YAML(typ="rt")
    yaml.preserve_quotes = True
    yaml.width = 4096  # never re-wrap an author's long line
    yaml.indent(mapping=2, sequence=4, offset=2)
    yaml.representer.add_representer(
        type(None),
        lambda dumper, _data: dumper.represent_scalar("tag:yaml.org,2002:null", "null"),
    )
    return yaml


def dump_round_trip(yaml: YAML, document: Any) -> str:
    """Serialize a round-trip document back to text."""
    buffer = StringIO()
    yaml.dump(document, buffer)
    return buffer.getvalue()


def write_artifact_text(path: Path, text: str) -> None:
    """Write an artifact back: atomically, and without touching its line endings.

    The one write path for every pass that rewrites an artifact (repair, conform, the
    combined pipeline), so the "atomic, endings untouched" contract has a single home.
    ``newline=""`` disables Python's newline translation — the text carries exactly the
    endings the caller preserved.
    """
    from strif import atomic_output_file

    with atomic_output_file(path) as tmp:
        Path(tmp).write_text(text, encoding="utf-8", newline="")
